### PR TITLE
fix: [DHIS2-10109] Filter out program rules for data elements that are not supposed to be applied

### PR DESCRIPTION
**Pull request by enricocolasante**

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/programrule/ShowErrorWarningImplementerTest.java [hunk 1]
```diff
@@ -71,173 +63,133 @@
...
 
-        ProgramRuleAction programRuleActionErrorOnComplete = createProgramRuleAction( 'B', programRule );
-        programRuleActionErrorOnComplete.setProgramRuleActionType( ProgramRuleActionType.ERRORONCOMPLETE );
+    private TrackerBundle bundle;
 
-        programRuleActionErrorOnComplete.setContent( CONTENT );
-        programRuleActionErrorOnComplete.setData( DATA );
-        programRuleActionService.addProgramRuleAction( programRuleActionErrorOnComplete );
+    @Mock
+    private TrackerPreheat preheat;
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/programrule/ShowErrorWarningImplementerTest.java:97]: Mocking data is a bit controversial, in general, it should be avoided. I guess in this case, it's not worth to change.

**Review by luciano-fiandesio — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/programrule/ShowErrorWarningImplementerTest.java (hunk 1 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/programrule/ShowErrorWarningImplementerTest.java:97]: When we test and there is the Preheat involved is always difficult to go to the actual service to mock it.
In the end, we are using the Preheat as a service that give us the data from the DB in a performant way.

Comment (luciano-fiandesio) [reply dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/programrule/ShowErrorWarningImplementerTest.java:97]: Sorry Enrico, I placed the cursor on the wrong place. I meant this:

```
@Mock
private ProgramStage programStage;
```

Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/programrule/ShowErrorWarningImplementerTest.java:97]: I agree, fixed

**Review by gnespolino — APPROVED** [review]

**Pull request merged**